### DISCUSSION TITLE
Streamlink api questions

### DISCUSSION THREAD

**Discussion by CrazyWolf13** [streamlink/streamlink#6049] [Q&A] [answered]
### Description

Hi

I kinda got lost a bit in the huge docs, and just tried my best at implementing it in python.

Is my current code correctly using streamlink?

-----

How may I use the --disable-ads here?

I tried with:
```python
    download_task.block_ads = True
    options = Options()
    options.set('disable_ads', download_task.block_ads)
    print(options.get('disable_ads'))
```
It returns True, however I still see the pre-roll screen ?

-----

Also I get quite some of these warnings:
```
Encountered a stream discontinuity. This is unsupported and will result in incoherent output data.
```

And also my output videos are sometimes barely playable, show incorrect times, video freezes at a 1-2 second interval...

-----

I've searched quite a bit before asking here, so I hope you can help me a bit.

https://streamlink.github.io/api/options.html#streamlink.options.Options

This seems like the correct page, but I'm kinda confused between implementing a new argument, which is not what I want, and just setting an argument.

Or do I need to import this:
https://github.com/streamlink/streamlink/blob/6.7.4/src/streamlink/plugins/twitch.py#L885

And the use _get_streams function?
---

My current code:

 ```python
    session = streamlink.Streamlink()
    try:
        plugin_name, plugin_class, _ = session.resolve_url(url)
    except streamlink.NoPluginError:
        raise HTTPException(status_code=404, detail="No plugin found for the given URL.")

    streams = session.streams(url)
    if not streams:
        raise HTTPException(status_code=404, detail="No streams found.")

    stream = streams.get(quality)
    if not stream:
        raise HTTPException(status_code=400, detail="Desired stream quality not found.")

    with open(./Video.mp4, "wb") as f:
        stream_fd = stream.open()
        try:
            while True:
                data = stream_fd.read(1024)
                if not data:
                    break
                f.write(data)
        finally:
            stream_fd.close()
```

Thanks!

**Comment by bastimeyer** [comment]
> Is my current code correctly using streamlink?

No. There's redundant stuff that doesn't make sense.

Read the docs of the `Streamlink.streams()` method, which resolves / looks up matching plugins from the input URL, instantiates the first matching `Plugin` and immediately resolves the streams from the input URL. You are trying to do both for some reason, first resolving a plugin which you then don't use and then you're calling `Streamlink.streams()`. Just call `Plugin.streams()` in order to resolve the streams from the resolved plugin.

- https://streamlink.github.io/api/session.html#streamlink.session.Streamlink.streams
- https://streamlink.github.io/api/plugin.html#streamlink.plugin.Plugin
- https://streamlink.github.io/api/plugin.html#streamlink.plugin.Plugin.streams

> How may I use the --disable-ads here?
> It returns True, however I still see the pre-roll screen ?

There's no `--disable-ads` CLI argument. CLI arguments are also not a 1:1 mapping to session or plugin option names.

You want to set the `disable-ads` Twitch plugin option, which is not a global session option. Just creating an `Options` object doesn't do anything on its own, so what you're attempting to do doesn't make sense. You need to pass it to the Twitch plugin constructor (after you have resolved the plugin class) or to `Streamlink.streams()`. Any global options like HTTP settings or HLS stream settings need to be set on the `Streamlink` session instance however.

- https://streamlink.github.io/api/session.html#streamlink.session.Streamlink.set_option
- https://streamlink.github.io/api/plugin.html#streamlink.plugin.Plugin

> Also I get quite some of these warnings:

See #6022, which was just merged a few hours ago. However, regardless whether you let Streamlink filter out ads on Twitch or not, there will be stream discontinuities on Twitch.

> ```python
> while True:
>     data = stream_fd.read(1024)
>     if not data:
>         break
>     f.write(data)
> ```

This will block your main thread, which is pretty bad.

**Comment by CrazyWolf13** [comment]
Hi 

Thanks for the reply, yeah you're right!

Altough I really like the in-depht documentation, it's quite overwhelming for junior devs.

I'm unable to get anything working, could you please share some basic examples?

That'd porbably make everyone's life easier :)

Here my current code, where I tried to apply the things you mentioned:

Also I currently have the whole streamlink async function called inside asyncio, is that still a problem then?

```python
import streamlink

url = "https://twitch.tv/LetsHugoTV"
stream_quality = "160p"

session = streamlink.Streamlink()

try:
    plugin = session.resolve_url(url)
    plugin.options.set("disable_ads", True) 
    streamlink.stream(url, stream_quality)
    
    streams = streamlink.streams()
    
    with open("./video.mp4", "wb") as f:
        stream_fd = streamlink.open()
        try:
            while True:
                data = stream_fd.read(1024)
                if not data:
                    break
                f.write(data)
        finally:
            stream_fd.close()

except streamlink.exceptions.NoStreamsError:
    print("Could not find the desired quality.")
except streamlink.exceptions.NoPluginError:
    print("Could not find a plugin for the given URL.")
except streamlink.exceptions.PluginError as e:
    print(f"Plugin error: {e}")
except Exception as e:
    print(f"An error occurred: {e}")
```

Thanks for any help / guidance!

**Reply by bastimeyer** [reply]
Sorry, but I'm not here to write any code for you. You'll have to figure this out on your own. I've already explained what you were doing wrong. The documentation and source code is freely available for anyone to study.

> Also I currently have the whole streamlink async function called inside asyncio, is that still a problem then?

Streamlink does not provide an async API, so when using async code while reading the stream like you're doing in the code above, this won't protect you from blocking the thread.

**Comment by CrazyWolf13** [comment, marked as answer]
> Sorry, but I'm not here to write any code for you.

Yeah sure no problem with that, I did also not expect that, just maybe an example, on a general config on this topic, because looking at the discussions, I was not the only one having trouble to understand (Interpret the docs).

---

For anyone else, here is a working example with plugin config and global configs added.

```python
    session = Streamlink()
    options = Options()
    # Plugin Setting
    options.set("disable_ads", block_ads)
    # Global / Session Options
    session.set_option("stream-timeout", 30)
   # Don't forget to append the actual options list here
    streams = session.streams(url, options)

    with open("./video.mp4", "wb") as f:
        fd = streams[stream_quality].open()
        try:
            while True:
                data = fd.read(1024)
                if not data:
                    break
                f.write(data)
        finally:
            fd.close()
```

Here some ressources that have been the most helpful to me:
https://streamlink.github.io/api_guide/quickstart.html -> This was the most helpful to me, have never seen that before.
https://streamlink.github.io/migrations.html#streamlink-g-s-et-plugin-option -> This explains the part with the options and has some example, examples are crucial.
https://github.com/streamlink/streamlink/discussions/5713
https://github.com/streamlink/streamlink/discussions/6027
